task 1/main.cpp

```diff
@@ -0,0 +1,267 @@
+    MyPoint(double x = 0.0, double y = 0.0): x(x), y(y) { }   //构造函数  
+  
+    friend istream& operator >> (istream& in, MyPoint& P) { return in >> P.x >> P.y; }  
+    friend ostream& operator << (ostream& out, const MyPoint& P) { return out << P.x << ' ' << P.y; }  
+  
+    friend Vector operator + (const Vector& A, const Vector& B) { return Vector(A.x+B.x, A.y+B.y); }  
+    friend Vector operator - (const MyPoint& A, const MyPoint& B) { return Vector(A.x-B.x, A.y-B.y); }  
+    friend Vector operator * (const Vector& A, const double& p) { return Vector(A.x*p, A.y*p); }  
+    friend Vector operator / (const Vector& A, const double& p) { return Vector(A.x/p, A.y/p); }  
+    friend bool operator == (const MyPoint& A, const MyPoint& B) { return dcmp(A.x-B.x) == 0 && dcmp(A.y-B.y) == 0; }  
+    friend bool operator < (const MyPoint& A, const MyPoint& B) { return A.x < B.x || (A.x == B.x && A.y < B.y); }  
+  
+    void in(void) { scanf("%lf%lf", &x, &y); }  
+    void out(void) { printf("%lf %lf", x, y); }  
+};  
+  
+#define PI 3.14159265
+const double ConvertToJiaodu=180.0 / PI;
+double MyDot(const Vector& A, const Vector& B) { return A.x*B.x + A.y*B.y; }  //点积  
+double MyLength(const Vector& A){ return sqrt(MyDot(A, A)); }  
+double MyAngle(const Vector& A, const Vector& B) { return acos(MyDot(A, B)/MyLength(A)/MyLength(B))*ConvertToJiaodu; }  //向量夹角  
+/*
+输入为一个旋转矩形
+输出为一个向量（计算几何模板中定义），代表该旋转矩形长轴的方向，仅代表直线方向，向量正负没有实际意义
+
+*/
+Vector mainVecOfRect(RotatedRect& r1)
+{
+    Point2f vers[4];//四个顶点
+    r1.points(vers);
+    Vector xs[2]={{0,0},{0,0}};//存储两个向量，比较长度
+    for(int i=0;i<=1;i++)
+    {
+        MyPoint ds(vers[i].x,vers[i].y);
+        MyPoint ed(vers[i+1].x,vers[i+1].y);
+        Vector biu(ds.x-ed.x,ds.y-ed.y);
+        xs[i]=biu;
+    }
+    if(MyLength(xs[0])>MyLength(xs[1]))
+    return xs[0];
+    else
+    return xs[1];
+}
+/*
+输入：已经判定为灯条的两个旋转矩形轮廓
+输出：是否是配对的灯条
+*/
+bool pairing(RotatedRect& x1,RotatedRect& x2)
+{
+    double deltaangle=abs(x1.angle-x2.angle);//旋转角度差，如果相近可能是方向垂直或者平行
+    
+   if(deltaangle<1.8)
+   {
+       Vector s1=mainVecOfRect(x1),s2=mainVecOfRect(x2);
+       //如果长轴互相垂直的话，不配对
+       if(fabs((MyAngle(s1,s2)-90))<5.0)
+       return false;
+        Vector zhou(x1.center.x-x2.center.x,x1.center.y-x2.center.y);
+        //如果两个矩形中心连线并不垂直于他们的长轴的话，那么不是矩形的轮廓，也不配对
+        if(fabs(MyAngle(zhou,s1)-90)>10)
+        return false;
+       return true;
+   }
+   else
+   {
+       return false;
+   }
+   
+}
+/*
+输入：一个RGB通道彩色图片，绘画颜色，需要绘画的旋转矩形，绘画的线条长度
+在目标图片上绘画目标颜色的矩形
+*/
+void StrokeRectOnMat(Mat& des,Scalar Color,RotatedRect tar,int thick)
+{
+     Point2f vers[4];
+      tar.points(vers);
+      for(int j=0;j<4;j++)
+             {
+            line(des,vers[j],vers[(j+1)%4],Color,thick);       
+             }
+        
+}
+/*
+输入：两个坐标点
+输出：两点连线的中点的坐标
+*/
+MyPoint middlePoint(MyPoint xxx,MyPoint yyy)
+{
+    Vector biu(xxx.x-yyy.x,xxx.y-yyy.y);
```

Comment:
please use some more informative names for your variables.

Reply:
Sorry, this part has some self releasing and I forgot to correct it ,  I won't do it in the next time, Sorry!